numpy/ma/core.py

```diff
@@ -3233,7 +3216,29 @@ def __setitem__(self, indx, value):
                 _mask[indx] = mval
         elif not self._hardmask:
             # Unshare the mask if necessary to avoid propagation
+            # We want to remove the unshare logic in the future, when this
+            # is removed, the unshare_mask function should likely be deprecated
             if not self._isfield:
+                if self._sharedmask and not (
+                        # No one else _mask, self._mask + 1 holds a reference
+                        # and the array owns its own data
+                        (sys.getrefcount(_mask) == 3) and _mask.flags.owndata):
+                    # 2016.01.15 -- v1.11.0
+                    warnings.warn(
+                       "setting an item on a masked array which has a shared "
```

Comment:
I tried to do it (with release notes changes), but I am not the best writer for this stuff....